components/datetime/src/format/mod.rs

```diff
@@ -143,42 +146,18 @@ where
                 FieldSymbol::Second(..) => {
                     format_number(w, date_time.second().into(), field.length)?
                 }
-                FieldSymbol::DayPeriod(period) => match period {
-                    fields::DayPeriod::AmPm => {
-                        let symbol =
-                            data.get_symbol_for_day_period(period, field.length, date_time.hour());
-                        w.write_str(symbol)?
-                    }
-                },
+                FieldSymbol::DayPeriod(period) => {
+                    let symbol = data.get_symbol_for_day_period(
+                        period,
+                        field.length,
+                        date_time.hour(),
+                        date_time.minute(),
+                    );
+                    w.write_str(symbol)?
+                }
             },
             PatternItem::Literal(l) => w.write_str(l)?,
         }
     }
     Ok(())
 }
-
-#[cfg(test)]
-mod tests {
-    use super::*;
-
-    #[test]
-    fn test_format_numer() {
-        let values = &[2, 20, 201, 2017, 20173];
-        let samples = &[
-            (FieldLength::One, ["2", "20", "201", "2017", "20173"]),
-            (FieldLength::TwoDigit, ["02", "20", "01", "17", "73"]),
-            (
-                FieldLength::Abbreviated,
-                ["002", "020", "201", "2017", "20173"],
-            ),
-            (FieldLength::Wide, ["0002", "0020", "0201", "2017", "20173"]),
-        ];
-        for (length, expected) in samples {
-            for (value, expected) in values.iter().zip(expected) {
-                let mut s = String::new();
-                format_number(&mut s, *value, *length).unwrap();
-                assert_eq!(s, *expected);
-            }
-        }
-    }
-}
```

Comment:
This test moved to the new file in `tests/mod.rs`

Reply:
Is this a conventional place to save tests? @zbraniecki @Manishearth

Reply:
Why not */components/datetime/tests/dayperiod.rs* ?

Reply:
Given that you are writing these tests as parameterized tests with fixtures, I definitely think that they belong as an integration test, not hidden in a test directory in the library.

(sorry for the triple-comment)

Reply:
I would have liked for it to be an integration test, but as I noted in the description of the PR, there is no public functionality to format dates with a custom pattern, so it has to live within the  private`format` module. This seemed, to me, to be the best compromise, but I'm open to other suggestions.

Reply:
Hmm.  This is an interesting question.  You could, in principle, do this as an integration test with a custom data provider that feeds in your custom pattern strings.  That should actually be pretty easy; see [StructProvider](https://unicode-org.github.io/icu4x-docs/doc/icu_provider/struct_provider/struct.StructProvider.html).  Let's discuss.

Reply:
For unit tests it's fine to create a `tests` module. But if this can be an integration test, that may be better.

Reply:
@sffc So I looked into this a bit more this morning, specifically into how I might use `StructProvider`.

Even if I used `StructProvider` to get a custom payload, it seems that both [`write_pattern()`](https://github.com/unicode-org/icu4x/blob/273ec285057bad902c6a08880df4c07a188a7968/components/datetime/src/format.rs#L91) and [`DateTimeFormat`](https://github.com/unicode-org/icu4x/blob/273ec285057bad902c6a08880df4c07a188a7968/components/datetime/src/lib.rs#L112) must take a [`DatesV1`](https://github.com/unicode-org/icu4x/blob/273ec285057bad902c6a08880df4c07a188a7968/components/provider/src/structs/dates.rs#L14) struct specifically as their data. 

Unless there is another way to format the datetime using the custom provider, I'm not seeing any viable options to format with a custom pattern publicly. That being said, I'm still trying to build an overall familiarity with the code base, so I certainly could have missed a key insight here.

Reply:
Right.  Create a DatesV1, either using `Default::default()` or by loading a stock locale from test data, then modify a field like `patterns.time.full` with your desired pattern.

I like the idea of data-driven tests being able to overwrite locale data for that testing.  @zbraniecki  thoughts?

Reply:
Okay, so I've now moved this test back to the file where it existed previously, and I moved the new tests to the `components/datetime/tests` folder to be integration tests. I really like this a lot more. 

But it feels somewhat hacky to me to grab the CLDR data for the locale, and then just overwrite the `long` patterns for each relevant case in my test data. Is this really intended functionality? I had previously overlooked that every member in `DatesV1` is public. 

I had a conversation with @zbraniecki about how custom patterns are often misused in practice. My understanding is that this is why we don't really have a public functionality to directly do something like `DateTimeFormat::try_format(&date_time, &pattern)`, and that we prefer to have people create a `DateTimeFormat` using the options bags. 

It just seems strange to me to have this preference to funnel people through pre-existing options, but still allow "backdoors" for custom patterns by letting people directly mutate the data. I'm almost of the opinion that, if the functionality exists anyway, why not provide something like`DateTimeFormat::try_format(&date_time, &pattern)`?

I suppose I just want to gain clarity around the design decisions behind _not_ having a custom format with pattern function, and for making all of the `DatesV1` members public, since I wasn't involved in the project at the time when these were conceived or implemented.

In the interest of not having duplicated conversations, if there are links to relevant pre-existing conversations that would be helpful to me.

Reply:
> In the interest of not having duplicated conversations, if there are links to relevant pre-existing conversations that would be helpful to me.

We've had conversations about this before, but I think they weren't written down.

The fields in all the data structs are public by design. The data struct is designed to be the place where you can "overlay" stock data with custom overrides.

That being said, I agree that in this particular case, it would probably be cleaner to have a low level pattern API. We can revisit this test when we have such API.